sklearn/feature_selection/_from_model.py

```diff
@@ -260,18 +286,23 @@ def fit(self, X, y=None, **fit_params):
             Fitted estimator.
         """
         if self.max_features is not None:
-            if not isinstance(self.max_features, numbers.Integral):
-                raise TypeError(
-                    "'max_features' should be an integer between"
-                    " 0 and {} features. Got {!r} instead.".format(
-                        X.shape[1], self.max_features
+            if isinstance(self.max_features, numbers.Integral):
+                if self.max_features < 0 or self.max_features > X.shape[1]:
+                    raise ValueError(
+                        "When using an int, 'max_features' should be"
+                        f" between 0 and {X.shape[1]}. Got {self.max_features} instead."
                     )
-                )
-            elif self.max_features < 0 or self.max_features > X.shape[1]:
-                raise ValueError(
-                    "'max_features' should be 0 and {} features.Got {} instead.".format(
-                        X.shape[1], self.max_features
+                self.max_features_ = self.max_features
+            elif callable(self.max_features):
+                self.max_features_ = self.max_features(X)
+                if not isinstance(self.max_features_, numbers.Integral):
+                    raise ValueError(
+                        "When `max_features` is a callable, it must return an int"
                     )
```

Comment:
you could maybe refactor the code here, first have a local `max_features` to be either `self.max_features` or `self.max_features(X)`, and then run `check_scalar` on it, then set the `max_features_` attribute.